# Exploring eBay Car Sales Data

In this project, we'll work with dataset of used cars from eBay Kleinanzeigen. We have the modified version of [this dataset](https://data.world/data-society/used-cars-data). It's dirtier than the original because we want to simulate a real data-cleaning task before data-analyze.

The dataset consists of next columns:

- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing
- `price` - The price on the ad to sell the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The vehicle Type.
- `yearOfRegistration` - The year in which the car was first registered.
- `gearbox` - The transmission type.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `odometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month in which the car was first registered.
- `fuelType` - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage` - If the car has a damage which is not yet repaired.
- `dateCreated` - The date on which the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeenOnline` - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
import numpy as np
import pandas as pd

autos = pd.read_csv('autos.csv', encoding="Latin-1")

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
print(autos.info())
print('\n')
print(autos.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

We can see next details after first glance of our data:
- The dataset includes 50k rows and 20 columns
- Most of the data are strings even data that we want to see as numbers (e.g. columns `price`, `odometer`)
- Some columns contain string data in German (`notRepairedDamage`, `fuelType`, `vehicleType`, `gearbox`, `model`).
- The date of registration was separated into two columns: `monthOfRegistration` and `yearOfRegistration`
- `name` column is not in human-readable format
- Some columns contain null values, but not more than ~20% of the data per column

## Cleaning Column Names

The column names in the dataset are in camelCase format, but in Python, we prefer snake_case. So let's convert column names to snake_case. We will use [rename method](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rename.html) to do it.

In [4]:
autos.rename(columns={ "dateCrawled": "date_crawled", "offerType": "offer_type", "abtest": "ab_test", "vehicleType": "vehicle_type", "yearOfRegistration": "registration_year", "powerPS": "power_ps", "monthOfRegistration": "registration_month", "fuelType": "fuel_type", "notRepairedDamage": "unrepaired_damage", "dateCreated": "ad_created", "nrOfPictures": "num_photos", "postalCode": "postal_code", "lastSeen": "last_seen" }, inplace=True)

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

In addition to the simple conversion from camelCase to snake_case, we also renamed several columns for a better description (e.g. `nrOfPictures` to `num_photos` or `dateCreated` to `ad_created`).

## Initial Exploration and Cleaning

Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for:
- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
- Examples of numeric data stored as text which can be cleaned and converted.

In [5]:
description = autos.describe(include="all")
description

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-14 20:50:02  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

Let's investigate columns with unique row values equal NaN

In [6]:
unique = description.loc['unique']

# use boolean indexing to select only the columns with NaN values
nan_columns = unique.loc[unique.isna()]
nan_columns

registration_year     NaN
power_ps              NaN
registration_month    NaN
num_photos            NaN
postal_code           NaN
Name: unique, dtype: object

In [7]:
autos['registration_year'].value_counts()

2000    3354
2005    3015
1999    3000
2004    2737
2003    2727
        ... 
1931       1
1929       1
1001       1
9996       1
1952       1
Name: registration_year, Length: 97, dtype: int64

In [8]:
autos['power_ps'].value_counts()

0        5500
75       3171
60       2195
150      2046
140      1884
         ... 
650         1
490         1
362         1
153         1
16312       1
Name: power_ps, Length: 448, dtype: int64

In [9]:
autos['registration_month'].value_counts()

0     5075
3     5071
6     4368
5     4107
4     4102
7     3949
10    3651
12    3447
9     3389
11    3360
1     3282
8     3191
2     3008
Name: registration_month, dtype: int64

In [10]:
autos['num_photos'].value_counts()

0    50000
Name: num_photos, dtype: int64

In [11]:
autos['postal_code'].value_counts()

10115    109
65428    104
66333     54
45888     50
44145     48
        ... 
23942      1
83365      1
95683      1
97794      1
67585      1
Name: postal_code, Length: 7014, dtype: int64

Conclusions after researching columns:
- In `registration_month` column there are values of `0` and `12`. This is incorrect, because we can have months ranging from `0 to 11` or `1 to 12`.
- In `power_ps` column there is abnormal value `16312`
- `num_photos` column consists of zero values
- There are a number of text columns where all (or nearly all) of the values are the same: `seller` and `offer_type`

Let's drop `num_photos`, `seller` and `offer_type` columns

In [12]:
autos.drop(['num_photos', 'seller', 'offer_type'], axis=1, inplace=True)

Also, we can see that `price` and `odometer` columns contain string values instead of numeric ones. Let's convert it to numeric values.

In [13]:
def convert_price(price):
    return int(price.replace('$', '').replace(',', ''))

autos['price'] = autos['price'].apply(convert_price)
autos['price'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [14]:
def convert_odometer(km):
    return int(km.replace(',', '').replace('km', ''))

autos['odometer'] = autos['odometer'].apply(convert_odometer)
autos.rename(columns={ 'odometer': 'odometer_km' }, inplace=True)
autos['odometer_km'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

## Exploring the Odometer and Price Columns

Let's start with `price` column

In [15]:
autos['price'].value_counts().sort_index()

0           1421
1            156
2              3
3              1
5              2
            ... 
10000000       1
11111111       2
12345678       3
27322222       1
99999999       1
Name: price, Length: 2357, dtype: int64

We can see some abnormal values: `0` and very big numbers like as `99999999`. I think that cars with zero prices are cars with unknown prices, but prices from $1 and up are normal prices because eBay is auction. Let's clean our data and leave only cars with prices from `$1` to `$100,000`. This price range looks most realistic for eBay.

In [16]:
autos = autos[autos['price'].between(1,100000)]
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price  ab_test vehicle_type  registration_year    gearbox  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

In [17]:
autos['price'].value_counts().sort_index()

1        156
2          3
3          1
5          2
8          1
        ... 
93911      1
94999      1
98500      1
99000      2
99900      2
Name: price, Length: 2309, dtype: int64

Let's investigate `odometer_km` column

In [18]:
autos['odometer_km'].value_counts().sort_index()

5000        827
10000       250
20000       754
30000       774
40000       814
50000      1010
60000      1153
70000      1217
80000      1414
90000      1734
100000     2113
125000     5055
150000    31411
Name: odometer_km, dtype: int64

Here all values look correct

## Exploring the date columns

In the dataset we have five columns with dates:
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

Let's see how these columns are presented in the dataset

In [19]:
autos[['date_crawled','ad_created','last_seen','registration_month','registration_year']]

date_crawled           ad_created            last_seen  \
0      2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54   
1      2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08   
2      2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37   
3      2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28   
4      2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50   
...                    ...                  ...                  ...   
49995  2016-03-27 14:38:19  2016-03-27 00:00:00  2016-04-01 13:47:40   
49996  2016-03-28 10:50:25  2016-03-28 00:00:00  2016-04-02 14:18:02   
49997  2016-04-02 14:44:48  2016-04-02 00:00:00  2016-04-04 11:47:27   
49998  2016-03-08 19:25:42  2016-03-08 00:00:00  2016-04-05 16:45:07   
49999  2016-03-14 00:42:12  2016-03-13 00:00:00  2016-04-06 21:18:48   

       registration_month  registration_year  
0                       3               2004  
1                       6               1997  
2                       7               2009  
3                       6               2007  
4                       7               2003  
...                   ...                ...  
49995                   1               2011  
49996                   5               1996  
49997                  11               2014  
49998                  11               2013  
49999                   1               1996  

[48526 rows x 5 columns]

`date_crawled`, `ad_created`, and `last_seen` are presented as string `YYYY-MM-DD HH:MM:SS`, `registration_month` and `registration_year` are presented as numbers.

Let's explore dates from crawler

In [20]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025306
2016-03-06    0.014054
2016-03-07    0.036043
2016-03-08    0.033281
2016-03-09    0.033096
2016-03-10    0.032210
2016-03-11    0.032601
2016-03-12    0.036929
2016-03-13    0.015682
2016-03-14    0.036578
2016-03-15    0.034270
2016-03-16    0.029613
2016-03-17    0.031633
2016-03-18    0.012880
2016-03-19    0.034806
2016-03-20    0.037897
2016-03-21    0.037361
2016-03-22    0.032910
2016-03-23    0.032230
2016-03-24    0.029345
2016-03-25    0.031612
2016-03-26    0.032189
2016-03-27    0.031076
2016-03-28    0.034847
2016-03-29    0.034126
2016-03-30    0.033652
2016-03-31    0.031839
2016-04-01    0.033652
2016-04-02    0.035507
2016-04-03    0.038598
2016-04-04    0.036517
2016-04-05    0.013086
2016-04-06    0.003174
2016-04-07    0.001401
Name: date_crawled, dtype: float64

We can see those ads were crawled during one month in 2016.

In [21]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038845
2016-04-04    0.036887
2016-04-05    0.011808
2016-04-06    0.003256
2016-04-07    0.001257
Name: ad_created, Length: 76, dtype: float64

The dataset contains ads created from 2015 to 2016.

In [22]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001072
2016-03-06    0.004328
2016-03-07    0.005399
2016-03-08    0.007419
2016-03-09    0.009582
2016-03-10    0.010675
2016-03-11    0.012385
2016-03-12    0.023802
2016-03-13    0.008902
2016-03-14    0.012612
2016-03-15    0.015888
2016-03-16    0.016445
2016-03-17    0.028109
2016-03-18    0.007336
2016-03-19    0.015847
2016-03-20    0.020649
2016-03-21    0.020628
2016-03-22    0.021391
2016-03-23    0.018526
2016-03-24    0.019742
2016-03-25    0.019227
2016-03-26    0.016754
2016-03-27    0.015641
2016-03-28    0.020855
2016-03-29    0.022359
2016-03-30    0.024791
2016-03-31    0.023802
2016-04-01    0.022813
2016-04-02    0.024914
2016-04-03    0.025203
2016-04-04    0.024502
2016-04-05    0.124634
2016-04-06    0.221840
2016-04-07    0.131929
Name: last_seen, dtype: float64

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [23]:
autos['registration_year'].describe()

count    48526.000000
mean      2004.754935
std         88.678603
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

We can see that in average cars were registered in 2006. Also, we can see two abnormal values: `min 1000` and `max 9999`.

## Dealing with Incorrect Registration Year Data

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

We'll take ads from the date range between 1950 and 2016.

In [24]:
autos = autos[autos['registration_year'].between(1950,2016)]

In [25]:
autos['registration_year'].value_counts(normalize=True).head(30)

2000    0.067672
2005    0.062975
1999    0.062138
2004    0.057977
2003    0.057891
2006    0.057269
2001    0.056518
2002    0.053322
1998    0.050620
2007    0.048818
2008    0.047467
2009    0.044721
1997    0.041826
2011    0.034790
2010    0.033997
1996    0.029450
2012    0.028034
1995    0.026297
2016    0.026039
2013    0.017138
2014    0.014178
1994    0.013491
1993    0.009116
2015    0.008322
1992    0.007915
1990    0.007443
1991    0.007271
1989    0.003732
1988    0.002874
1985    0.002038
Name: registration_year, dtype: float64

We can see that most of the cars were registered in the early 2000s.

## Exploring Price by Brand

In [26]:
brands = autos['brand'].value_counts(normalize=True)
brands

volkswagen        0.211510
bmw               0.110098
opel              0.107653
mercedes_benz     0.096478
audi              0.086654
ford              0.069903
renault           0.047188
peugeot           0.029879
fiat              0.025675
seat              0.018296
skoda             0.016430
nissan            0.015293
mazda             0.015207
smart             0.014178
citroen           0.014028
toyota            0.012719
hyundai           0.010038
sonstige_autos    0.009545
volvo             0.009159
mini              0.008773
mitsubishi        0.008236
honda             0.007850
kia               0.007078
alfa_romeo        0.006649
suzuki            0.005941
chevrolet         0.005705
porsche           0.005577
chrysler          0.003518
dacia             0.002638
daihatsu          0.002510
jeep              0.002274
subaru            0.002145
land_rover        0.002102
saab              0.001652
jaguar            0.001566
daewoo            0.001501
trabant           0.001373
r

We can see that `volkswagen`, `bmw`, and `opel` take up almost 50% of all ads. We'll explore brands with `>5%` of the total values.

In [27]:
most_popular_brands = brands[brands > .05].index
most_popular_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

Let's find the average price for cars of these brands

In [28]:
mean_price_by_brands = {}

for brand in most_popular_brands:
    mean_price_by_brands[brand] = int(autos[autos['brand'] == brand]['price'].mean())
    
mean_price_by_brands

{'volkswagen': 5398,
 'bmw': 8200,
 'opel': 2976,
 'mercedes_benz': 8526,
 'audi': 9295,
 'ford': 3698}

We can see that the most expensive brand is `audi`. It has average price is **9295**. The cheapest brand is `opel` with an average price is **2976**.

## Storing Aggregate Data in a DataFrame

For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price.

In [29]:
mean_mileage_by_brands = {}

for brand in most_popular_brands:
    mean_mileage_by_brands[brand] = int(autos[autos['brand'] == brand]['odometer_km'].mean())
    
mean_mileage_by_brands

{'volkswagen': 128710,
 'bmw': 132658,
 'opel': 129336,
 'mercedes_benz': 130894,
 'audi': 129188,
 'ford': 124397}

Let's convert both dictionaries to series objects, using the series constructor

In [30]:
mean_price_by_brands_series = pd.Series(mean_price_by_brands)
mean_price_by_brands_series

volkswagen       5398
bmw              8200
opel             2976
mercedes_benz    8526
audi             9295
ford             3698
dtype: int64

In [31]:
mean_mileage_by_brands_series = pd.Series(mean_mileage_by_brands)
mean_mileage_by_brands_series

volkswagen       128710
bmw              132658
opel             129336
mercedes_benz    130894
audi             129188
ford             124397
dtype: int64

Now we'll create dataframes from these series

In [32]:
df = pd.DataFrame(mean_price_by_brands_series, columns=['mean_price'])
df

mean_price
volkswagen           5398
bmw                  8200
opel                 2976
mercedes_benz        8526
audi                 9295
ford                 3698

In [33]:
df['mean_mileage'] = mean_mileage_by_brands_series
df

mean_price  mean_mileage
volkswagen           5398        128710
bmw                  8200        132658
opel                 2976        129336
mercedes_benz        8526        130894
audi                 9295        129188
ford                 3698        124397

There is no correlation between mileage and price, since all cars on average have the same mileage (~128,000km).

## Translation of German Words to English

Since English is the most popular language, let's translate the data in some columns into English.

As we know columns `vehicle_type`, `gearbox`, `fuel_type`, and `unrepaired_damage` contain German words. Let's check unique values in each column.

In [34]:
print(autos['vehicle_type'].unique())
print(autos['gearbox'].unique())
print(autos['fuel_type'].unique())
print(autos['unrepaired_damage'].unique())

['bus' 'limousine' 'kleinwagen' 'kombi' nan 'coupe' 'suv' 'cabrio'
 'andere']
['manuell' 'automatik' nan]
['lpg' 'benzin' 'diesel' nan 'cng' 'hybrid' 'elektro' 'andere']
['nein' nan 'ja']


We can see that we have nan values in these columns. So our next steps:
1. Replace German words with English translation
2. Replace `nan` with `unknown`

In [35]:
vehicle_type_mapper = { 'bus': 'bus', 'limousine': 'limousine', 'kleinwagen': 'compact', 'kombi': 'combi', 'coupe': 'coupe', 'suv': 'suv', 'cabrio': 'cabrio', 'andere': 'other' }
gearbox_mapper = { 'manuell': 'manual', 'automatik': 'automatic' }
fuel_type_mapper = { 'lpg': 'lpg', 'benzin': 'gasoline', 'diesel': 'diesel', 'cng': 'cng', 'hybrid': 'hybrid', 'electro': 'electro', 'andere': 'other' }
unrepaired_damage_mapper = { 'nein': 'no', 'ja': 'yes' }

translation_map = { 'vehicle_type': vehicle_type_mapper, 'gearbox': gearbox_mapper, 'fuel_type': fuel_type_mapper, 'unrepaired_damage': unrepaired_damage_mapper }

for col in translation_map:
    autos[col] = autos[col].map(translation_map[col]).fillna('unknown')

Let's check results

In [36]:
autos['vehicle_type'].value_counts(dropna=False)

limousine    12596
compact      10584
combi         8929
bus           4031
cabrio        2998
unknown       2699
coupe         2436
suv           1962
other          387
Name: vehicle_type, dtype: int64

In [37]:
autos['unrepaired_damage'].value_counts(dropna=False)

no         33791
unknown     8297
yes         4534
Name: unrepaired_damage, dtype: int64

## Search for the Most Common Brand/Model Combinations

Let's find the most common make/model combinations to see which specific cars sell more often

In [38]:
unique_brands = autos['brand'].unique()
brand_model_combination = {}

for brand in unique_brands:
    brand_model_combination[brand] = autos[autos['brand'] == brand]['model'].describe()['top']
    
brand_model_combination

{'peugeot': '2_reihe',
 'bmw': '3er',
 'volkswagen': 'golf',
 'smart': 'fortwo',
 'ford': 'focus',
 'chrysler': 'andere',
 'seat': 'ibiza',
 'renault': 'twingo',
 'mercedes_benz': 'c_klasse',
 'audi': 'a4',
 'sonstige_autos': nan,
 'opel': 'corsa',
 'mazda': '3_reihe',
 'porsche': '911',
 'mini': 'cooper',
 'toyota': 'yaris',
 'dacia': 'sandero',
 'nissan': 'micra',
 'jeep': 'grand',
 'saab': 'andere',
 'volvo': 'v70',
 'mitsubishi': 'colt',
 'jaguar': 'andere',
 'fiat': 'punto',
 'skoda': 'octavia',
 'subaru': 'legacy',
 'kia': 'andere',
 'citroen': 'andere',
 'chevrolet': 'andere',
 'hyundai': 'i_reihe',
 'honda': 'civic',
 'daewoo': 'matiz',
 'suzuki': 'andere',
 'trabant': '601',
 'land_rover': 'freelander',
 'alfa_romeo': '156',
 'lada': 'niva',
 'rover': 'andere',
 'daihatsu': 'cuore',
 'lancia': 'ypsilon'}

Since eBay has the most advertisements for Volkswagen cars for sale, we can conclude that the most popular car for sale is the **Volkswagen Golf**.

## Checking Correlation Between Damage and Price

Let's check how much cheaper are cars with damage than their non-damaged counterparts

In [39]:
autos[autos['unrepaired_damage'] == 'yes']['price'].mean()

2222.3098808998675

In [40]:
autos[autos['unrepaired_damage'] == 'no']['price'].mean()

6982.830635376284

We can see that the average price of cars without damage is almost 4 times higher

## Conclusion

In this project we did the following work:
- Cleaned and prepared input data to analyze
- Analyzed the correlation between mileage and price
- Analyzed the correlation between damage and price
- Found the most popular brands and most popular model on eBay